lib/Sema/CSSimplify.cpp

```diff
@@ -6682,13 +6682,12 @@ performMemberLookup(ConstraintKind constraintKind, DeclNameRef memberName,
   // through optional types.
   //
   // FIXME: Unify with the above code path.
-  if (result.ViableCandidates.empty() &&
-      baseObjTy->is<AnyMetatypeType>() &&
+  if (baseObjTy->is<AnyMetatypeType>() &&
       constraintKind == ConstraintKind::UnresolvedValueMember) {
     if (auto objectType = instanceTy->getOptionalObjectType()) {
       // If we don't have a wrapped type yet, we can't look through the optional
       // type.
-      if (objectType->getAs<TypeVariableType>()) {
+      if (objectType->getAs<TypeVariableType>() && result.ViableCandidates.empty()) {
```

Comment:
I think this would break source compatibility when score of something found on `Wrapped` be better than score on `Optional` itself...

Reply:
Since we already score `Decl` better than `DeclViaUnwrappedOptional`, the score on `Wrapped` would have to outweigh the disadvantage from the `OverloadChoiceKind` mismatch. Do you think such situations are likely to occur in practice?

Is there any way to unconditionally prefer the non-optional choices, in the case where we've found multiple solutions?

Reply:
> Since we already score Decl better than DeclViaUnwrappedOptional

Solutions are first ranked on the score produced by the solver (https://github.com/apple/swift/blob/main/lib/Sema/CSRanking.cpp#L758) and then their declaration diff is compared if scores match. It's quite possible that choices could have different optionality of require other implicit conversion which would change the behavior here.

Reply:
@xedin Ah, thank you for pointing that out. The situation where we have identically-named static members on both `Optional<Wrapped>` and `Wrapped` where both properly type-check seems pretty unlikely to me, but if it's above the risk level for being considered source-compatible (as opposed to just a bug) then so be it.

Is there a way to add the `Wrapped` overloads but not skip over them whenever any `Optional` overloads successfully type check?

Reply:
One way to implement this without changing existing behavior would be to lift `DeclViaUnwrappedOptional` into the solver score to make sure regular declarations are always preferred over the ones found on wrapped type.